docs/proposals/20210525-cluster-class-and-managed-topologies.md

```diff
@@ -0,0 +1,433 @@
+---
+title: ClusterClass and managed topologies
+authors:
+  - "@srm09"
+  - "@vincepri"
+  - "@fabriziopandini"
+reviewers:
+  - "@vincepri"
+  - "@fabriziopandini"
+creation-date: 2021-05-25
+status: provisional
+---
+
+# ClusterClass and Managed Topologies
+
+## Table of Contents
+
+- [Title](#title)
+  - [Table of Contents](#table-of-contents)
+  - [Glossary](#glossary)
+      - [ClusterClass](#clusterclass)
+      - [Managed Cluster topology](#managed-cluster-topology)
+      - [Worker class](#worker-class)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+      - [Goals](#goals)
+      - [Prospective future Work](#prospective-future-work)
+  - [Proposal](#proposal)
+      - [User Stories](#user-stories)
+        - [Story 1 - Use ClusterClass to easily stamp clusters](#story-1---use-clusterclass-to-easily-stamp-clusters)
+        - [Story 2 - Easier UX for kubernetes version upgrades](#story-2---easier-ux-for-kubernetes-version-upgrades)
+        - [Story 3 - Easier UX for scaling workers nodes](#story-3---easier-ux-for-scaling-workers-nodes)
+      - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+        - [New API types](#new-api-types)
+            - [ClusterClass](#clusterclass-1)
+        - [Modification to existing API Types](#modification-to-existing-api-types)
+            - [Cluster](#cluster)
+        - [Validations](#validations)
+            - [ClusterClass](#clusterclass-2)
+            - [Cluster](#cluster-1)
+        - [Behaviors](#behaviors)
+            - [Create a new Cluster using ClusterClass object](#create-a-new-cluster-using-clusterclass-object)
+            - [Update an existing Cluster using ClusterClass](#update-an-existing-cluster-using-clusterclass)
+        - [Provider implementation](#provider-implementation)
+            - [For infrastructure providers](#for-infrastructure-providers)
+            - [For Control plane providers](#for-control-plane-providers)
+      - [Risks and Mitigations](#risks-and-mitigations)
+  - [Alternatives](#alternatives)
+  - [Upgrade Strategy](#upgrade-strategy)
+  - [Additional Details](#additional-details)
+      - [Test Plan [optional]](#test-plan-optional)
+      - [Graduation Criteria [optional]](#graduation-criteria-optional)
+      - [Version Skew Strategy [optional]](#version-skew-strategy-optional)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+### ClusterClass
+A collection of templates that define a topology (control plane and machine deployments) to be used in one or more Managed Clusters.
+
+### Managed Cluster topology
```

Comment:
Copying over @CecileRobertMichon 's comment from the doc:
I could see this term being very confusing for users with "managed" (AKS, EKS, GKE) vs. "self-managed" clusters, externally "managed" infrastructure, and "management" clusters in CAPI terminology... managed is already a very overloaded term. Trying to think of alternatives.

Reply:
HUGE +1 from me as well. We really can't start referring to clusters built/defined with this new `ClusterClass` configuration by a different name (e.g., "managed cluster" vs "cluster"). In this particular case, the term "managed cluster" is too general, and in the cluster-api ecosystem it is very sensible right now to refer to *any* cluster-api workload cluster as a "managed cluster" — these cluster *are managed* by the cluster-api _management cluster_.

And, more broadly, this seems to be a natural evolution of the cluster-api abstraction. We should expect there to be many more in the future. We don't want to back ourselves into a corner by having to create a distinct terminology for every evolving "type" of cluster-api cluster configuration.

My 2¢ is that we should try to always refer to this in language as "a cluster that uses ClusterClass" to reflect that this is just one possible configuration paradigm for a cluster-api cluster, one that reflects a certain vintage ("I built my cluster from cluster-api version XYZ and am taking advantage of the latest functionality!") and operational preference. Rather than "a discrete _type_ of cluster-api cluster".

Reply:
moved conversation to https://github.com/kubernetes-sigs/cluster-api/pull/4678#discussion_r644679021